src/gg20/crypto_tools/paillier/zk/composite_dlog.rs

```diff
@@ -80,51 +89,80 @@ fn jacobi_symbol(a: &BigNumber, p: &BigNumber, q: &BigNumber) -> i32 {
 impl CompositeDLogStmt {
     #[allow(clippy::many_single_char_names, non_snake_case)]
     /// Setup the statement for Composite Dlog proof using the modulus `N = pq`.
-    /// This will generate an asymmetric basis `g` and a witness `s`.
+    /// This will generate an asymmetric basis `g` and a witness `s` such that
+    /// g^(-s) is also asymmetric basis.
     pub fn setup(
         rng: &mut (impl CryptoRng + RngCore),
         n: &BigNumber,
         p: &BigNumber,
         q: &BigNumber,
-    ) -> (Self, Randomness) {
-        // Sample an asymmetric basis g
-        let g = loop {
+        totient: &BigNumber,
+    ) -> (Self, CompositeDLogWitness, CompositeDLogWitness) {
+        let S = BigNumber::one() << WITNESS_SIZE;
+
+        loop {
+            // Sample an asymmetric basis g
             let g = BigNumber::random_with_rng(rng, n);
 
             // g is asymmetric when Jacobi symbol (g | n) = -1
-            if jacobi_symbol(&g, p, q) == -1 {
-                break g;
+            if jacobi_symbol(&g, p, q) != -1 {
+                continue;
             }
-        };
 
-        let S = BigNumber::one() << WITNESS_SIZE;
-        let s = Randomness::generate_with_rng(rng, &S);
-        let neg_s = Randomness(-&s.0);
+            // Sample s from {0,..,S-1} which is in Z*_N~ with high probability
+            let s = Randomness::generate_with_rng(rng, &S);
+            let neg_s = Randomness(-&s.0);
 
-        // v = g^(-s) mod N
-        let v = g.modpow(&neg_s.0, n);
+            // v = g^(-s) mod N
+            let v = g.modpow(&neg_s.0, n);
 
-        (Self { n: n.clone(), g, v }, s)
+            // Check if v is asymmetric
+            if jacobi_symbol(&v, p, q) != -1 {
+                continue;
+            }
+
+            let s_inv = if let Some(x) = s.0.invert(totient) {
+                // s^-1 mod phi(N) is treated as being sampled from {0,..,2^MAX_N_SIZE}
+                // and needs to be masked using an appropriately long `r`
+                CompositeDLogWitness {
+                    s: Randomness(x),
+                    size: MODULUS_MAX_SIZE,
+                }
+            } else {
+                continue;
+            };
+
+            let s = CompositeDLogWitness {
+                s,
+                size: WITNESS_SIZE,
+            };
+
+            return (Self { n: n.clone(), g, v }, s, s_inv);
+        }
     }
 }
 
 impl NIZKStatement for CompositeDLogStmt {
-    type Witness = Randomness;
+    type Witness = CompositeDLogWitness;
     type Proof = CompositeDLogProof;
 
     #[allow(non_snake_case)]
     fn prove(&self, wit: &Self::Witness, domain: &[u8]) -> Self::Proof {
-        let R = BigNumber::one() << R_SIZE;
-        let r = BigNumber::random(&R);
+        // Assume that v = g^(-s) mod N~
+        debug_assert!(self.v == self.g.modpow(&(-&wit.s.0), &self.n));
```

Comment:
Why only a debug assert for this?  I guess if we don't want to return an error then we can't guard against invalid witness.

Reply:
All our zkproofs currently assume that the statement and witness is already valid (which is the case currently). I have added similar `debug_asserts` only as a sanity check for testing. I would prefer to avoid calling more expensive operations like `modpow` as a sanity check in release mode.